planner/core/optimizer.go

```diff
@@ -187,6 +188,72 @@ func physicalOptimize(logic LogicalPlan) (PhysicalPlan, float64, error) {
 	return t.plan(), t.cost(), err
 }
 
+// eliminateUnionScanAndLock set lock property for PointGet and BatchPointGet and eliminates UnionScan and Lock.
+func eliminateUnionScanAndLock(sctx sessionctx.Context, p PhysicalPlan) PhysicalPlan {
+	var pointGet *PointGetPlan
+	var batchPointGet *BatchPointGetPlan
+	var physLock *PhysicalLock
+	var unionScan *PhysicalUnionScan
+	iteratePhysicalPlan(p, func(p PhysicalPlan) bool {
+		if len(p.Children()) > 1 {
+			return false
+		}
+		switch x := p.(type) {
+		case *PointGetPlan:
+			pointGet = x
+		case *BatchPointGetPlan:
+			batchPointGet = x
+		case *PhysicalLock:
+			physLock = x
+		case *PhysicalUnionScan:
+			unionScan = x
+		}
+		return true
+	})
```

Comment:
How about doing `transformPhysicalPlan` here to reduce plan tree iteration?

Reply:
transform is post-order, iterate is pre-order.
Use pre-order to iterate so we can stop early, and we can not do transform in pre-order.